lib/nexpose/node.rb

```diff
@@ -83,22 +83,28 @@ def method_missing(method, *args)
       # Finally the enumerations: names
       if method_name == 'names'
         @xml.xpath('./names/name').collect(&:text)
-
       elsif ['fingerprints', 'software'].include?(method_name)
-
-        xpath_selector = {
-          'fingerprints' => './fingerprints/os',
-          'software' => './software/fingerprint'
-        }[method_name]
-
-        xml_os = @xml.at_xpath(xpath_selector)
-        return '' if xml_os.nil?
-
-        xml_os.attributes['product'].value
+        get_fingerprint_product(method_name)
       else
         # nothing found, the tag is valid but not present in this ReportItem
         return nil
       end
     end
+
+    private
+
+    # returns the first 'product' value from the <fingerprints> or <software> element
+    def get_fingerprint_product(method_name)
+      xpath_selector = {
+        'fingerprints' => './fingerprints/os',
+        'software' => './software/fingerprint'
+      }[method_name]
+
+      xml_fingerprint = @xml.xpath(xpath_selector)
+      return '' if xml_fingerprint.nil?
+
+      product = xml_fingerprint.find { |fingerprint| fingerprint['product'] }
+      product ? product['product'] : 'n/a'
```

Comment:
final weirdness, if L106 is "filtering" to find elements w/ the right attribute, then why do we need to check again if we got it?

These two lines combined are doing some work but it's not very clear what work they're doing.

My mind also went to .fetch('product', 'n/a') or some other similar syntax rather than a conditional, but I don't know if that's supported.